src/aragonjs-wrapper.js

```diff
@@ -1,38 +1,166 @@
-import Aragon from '@aragon/wrapper'
-import { noop } from './utils'
+import Web3 from 'web3'
+import Aragon, {
+  providers,
+  setupTemplates,
+  isNameUsed,
+  ensResolve,
+} from '@aragon/wrapper'
+import { appDefaults, appIds, appLocator, ipfsDefaultConf } from './environment'
+import { noop, removeTrailingSlash } from './utils'
+import { getBlobUrl, WorkerSubscriptionPool } from './worker-utils'
 
 const ACCOUNTS_POLL_EVERY = 2000
 
-const IPFS_CONF_DEFAULT = {
-  rpc: { host: 'ipfs.infura.io', port: '5001', protocol: 'https' },
-}
-
-const appSrc = (app = {}) => {
+const appSrc = (app, gateway = ipfsDefaultConf.gateway) => {
   const hash = app.content && app.content.location
   if (!hash) return ''
 
-  // TODO: move this in the env settings
-  // This is the Voting app hash in the dev template
-  if (hash === 'QmV5sEjshcZ6mu6uFUhJkWM5nTa53wbHfRFDD4Qy2Yx88m') {
-    return 'http://localhost:3001/'
+  if (appLocator[app.appId]) {
+    return appLocator[app.appId]
   }
 
-  return `https://gateway.ipfs.io/ipfs/${hash}/`
+  return `${gateway}/${hash}/`
 }
 
 // Filter out apps without UI and add an appSrc property
-const prepareFrontendApps = apps =>
+const prepareFrontendApps = (apps, gateway) =>
   apps
-    .filter(app => app.content && app.name !== 'Vault')
-    .map(app => ({ ...app, appSrc: appSrc(app) }))
+    .map(app => ({ ...(appDefaults[app.appId] || {}), ...app }))
+    .filter(app => app && app['short_url'] && app.appId !== appIds['Vault'])
+    .map(app => ({ ...app, appSrc: appSrc(app, gateway) }))
+
+// Keep polling the main account.
+// See https://github.com/MetaMask/faq/blob/master/DEVELOPERS.md#ear-listening-for-selected-account-changes
+export const pollMainAccount = (provider = null, onAccount = noop) => {
+  const web3 = new Web3(provider)
+  let timer = -1
+  let lastFound = null
+  let stop = false
+
+  const poll = async () => {
+    const account = await getMainAccount(web3)
+    if (account !== lastFound) {
+      lastFound = account
+      onAccount(account)
+    }
+    if (stop) {
+      return
+    }
+    timer = setTimeout(poll, ACCOUNTS_POLL_EVERY)
+  }
+
+  poll()
+
+  return () => {
+    // In case we are waiting for getMainAccount
+    stop = true
+
+    clearTimeout(timer)
```

Comment:
Using a `setInterval` would make this cleaner; we wouldn't need the `stop` anymore. Something like:

```js
  const poll = async () => {
    const account = await getMainAccount(web3)
    if (account !== lastFound) {
      lastFound = account
      onAccount(account)
    }
  }
 
  poll()
  interval = setInterval(poll, ACCOUNTS_POLL_EVERY)
 
  return () => {
    clearInterval(timer)
  }
```

However, this isn't 100% robust, as we could be waiting for multiple `getMainAccount()`s when we clear the timer. A way around that would be to chain their promises and use `stop` as a poison pill:

```js
  const pollAccount = () => {
    return getMainAccount(web3).then(account => {
      if (account !== lastFound) {
        lastFound = account
        onAccount(account)
      }
    })
  }
 
  let lastPoll = pollAccount()
  interval = setInterval(
    () => {
      lastPoll = lastPoll.then(() => stop ? throw 'poison' : pollAccount())
    },
    ACCOUNTS_POLL_EVERY
  )
 
  return () => {
    // Poison pill any getMainAccounts we're waiting to execute
    stop = true
 
    clearInterval(timer)
  }
```

Reply:
> Using a setInterval would make this cleaner; we wouldn't need the stop anymore.

Using setInterval would mean that requests would be started every x time, no matter if the previous request is still being fetched, which could result in multiple requests being fetched at the same time. If the requests never end for some reason, it would also keep adding more and more requests.

```
[request ==]   [request =======]
                              [request =======]
[interval time][interval time][interval time]
```

Using setTimeout prevents this issue, and makes the interval time always the same between the end of a request and the start of the next one:

```
[request ==][interval time][request =======][interval time][request =======]
```

Also I think `if (stop) return` should be moved right after the `await`, to prevent any effect after the promise “cancellation”.

What do you think?

Reply:
Good points, what we currently have is strictly better than my second `clearInterval()` solution :).